src/bindings/fluid_ladspa.c

```diff
@@ -25,93 +25,103 @@
  * Author: Marcus Weseloh
  */
 
+#include "fluid_ladspa.h"
+
+#ifdef LADSPA
+
 #include "fluidsynth_priv.h"
 
-#include "fluid_ladspa.h"
 #include "fluid_sys.h"
-#include <dlfcn.h>
 #include <math.h>
 #include <ladspa.h>
 
-#define LADSPA_API_ENTER(_fx) (fluid_rec_mutex_lock((_fx)->api_mutex))
-
-#define LADSPA_API_RETURN(_fx, _ret)          \
-    fluid_rec_mutex_unlock((_fx)->api_mutex); \
-    return (_ret)
-
-#define FLUID_LADSPA_MAX_LIBS 100
-#define FLUID_LADSPA_MAX_PLUGINS 100
+#define FLUID_LADSPA_MAX_EFFECTS 100
 #define FLUID_LADSPA_MAX_NODES 100
-#define FLUID_LADSPA_MAX_PATH_LENGTH 512
 
 typedef enum _fluid_ladspa_state_t {
     FLUID_LADSPA_INACTIVE = 0,
     FLUID_LADSPA_ACTIVE,
     FLUID_LADSPA_RUNNING
-    
+
 } fluid_ladspa_state_t;
-    
-typedef enum _fluid_ladspa_node_type_t {
-    FLUID_LADSPA_NODE_AUDIO,
-    FLUID_LADSPA_NODE_CONTROL,
 
-} fluid_ladspa_node_type_t;
+typedef enum _fluid_ladspa_dir_t {
+    FLUID_LADSPA_INPUT,
+    FLUID_LADSPA_OUTPUT,
 
-typedef struct _fluid_ladspa_lib_t
-{
-    char *filename;
-    void *dlib;
-    LADSPA_Descriptor_Function descriptor;
+} fluid_ladspa_dir_t;
 
-} fluid_ladspa_lib_t;
+typedef enum _fluid_ladspa_node_type_t {
+    FLUID_LADSPA_NODE_AUDIO = 1,
+    FLUID_LADSPA_NODE_CONTROL = 2,
+    FLUID_LADSPA_NODE_EFFECT = 4,
+    FLUID_LADSPA_NODE_HOST = 8,
+    FLUID_LADSPA_NODE_USER = 16,
+
+} fluid_ladspa_node_type_t;
 
 typedef struct _fluid_ladspa_node_t
 {
     char *name;
     fluid_ladspa_node_type_t type;
-    LADSPA_Data *buf;
+
+    /* The buffer that LADSPA effects read and/or write to.
+     * If FluidSynth has been compiled WITH_FLOAT, then this
+     * points to the host buffer, otherwise it's a separate
+     * buffer. */
+    LADSPA_Data *effect_buffer;
+
+    /* Only set for host nodes, points to the host buffer */
+    fluid_real_t *host_buffer;
 
     int num_inputs;
     int num_outputs;
 
 } fluid_ladspa_node_t;
 
-typedef struct _fluid_ladspa_plugin_t
+typedef struct _fluid_ladspa_effect_t
 {
-    /* plugin instance id unique to the effects unit */
-    int id;
+    char *name;
 
+    /* Pointer to the library that this effect was loaded from */
+    fluid_module_t *lib;
+
+    /* The descriptor defines the plugin implementation, the
+     * handle points to an instance of that plugin */
     const LADSPA_Descriptor *desc;
     LADSPA_Handle *handle;
 
     int active;
 
+    /* Decides if we should call the run (mix = 0) or run_adding (mix = 1)
+     * plugin interface */
+    int mix;
+
     /* Used to keep track of the port connection state */
     fluid_ladspa_node_t **port_nodes;
 
-} fluid_ladspa_plugin_t;
+} fluid_ladspa_effect_t;
 
 struct _fluid_ladspa_fx_t
 {
     unsigned long sample_rate;
 
-    int audio_groups;
-    int effects_channels;
-    int audio_channels;
+    /* The buffer size for all audio buffers */
```

Comment:
"size for all audio buffers" is too unspecific, ie the unit is missing (bytes, audio frames?). Avoids tracing back it's usages to understand.